src/peergos/shared/io/ipfs/api/NamedStreamable.java

```diff
@@ -21,17 +24,35 @@
 
     class FileWrapper implements NamedStreamable {
         private final File source;
+        private final String pathPrefix;
 
-        public FileWrapper(File source) {
+        public FileWrapper(String pathPrefix, File source) {
             this.source = source;
+            this.pathPrefix = pathPrefix;
+        }
+
+        public FileWrapper(File source) {
+            this("", source);
         }
 
         public InputStream getInputStream() throws IOException {
             return new FileInputStream(source);
         }
 
+        public boolean isDirectory() {
+            return source.isDirectory();
+        }
+
+        public File getFile() {
+            return source;
+        }
+
         public Optional<String> getName() {
-            return Optional.of(source.getName());
+            try {
+                return Optional.of(URLEncoder.encode(pathPrefix + source.getName(), "UTF-8"));
```

Comment:
here for  example could  be ``remotePath.resolve(source.getName())`` if  pathPrefix was a ``java.nio.path.Path``